controllers/namespacescope/namespacescope.go

```diff
@@ -0,0 +1,122 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package nss
+
+import (
+	"context"
+	"time"
+
+	gset "github.com/deckarep/golang-set"
+	"k8s.io/apimachinery/pkg/types"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	nssv1 "github.com/IBM/ibm-namespace-scope-operator/api/v1"
+
+	"github.com/IBM/ibm-common-service-operator/controllers/bootstrap"
+)
+
+var (
+	apiGroupVersion = "operator.ibm.com/v1"
+	Kinds           = []string{"NamespaceScope"}
+	CRList          = []string{"common-service", "nss-odlm-scope"}
+	sourceCR        = "common-service"
+	targetCR        = "nss-odlm-scope"
+)
+
+var ctx = context.Background()
+
+// SyncUpCR syncs up the namespace members in source CR and target CR
+func SyncUpCR(bs *bootstrap.Bootstrap) {
+	for {
+		// wait for nss CRD
+		for _, kind := range Kinds {
+			if err := bs.WaitResourceReady(apiGroupVersion, kind); err != nil {
+				klog.Errorf("Failed to wait for resource ready with kind %s, apiGroupVersion: %s", kind, apiGroupVersion)
+				continue
+			}
+		}
+
+		// wait for source and target CR
+		for _, cr := range CRList {
+			for {
+				ready := waitCRReady(bs, cr, bs.CSData.MasterNs)
+				if ready {
+					break
+				}
+				time.Sleep(10 * time.Second)
+			}
+		}
+
+		// fetch the source and target NSS CR
+		sourceNsScope := &nssv1.NamespaceScope{}
+		sourceNsScopeKey := types.NamespacedName{Name: sourceCR, Namespace: bs.CSData.MasterNs}
+		if err := bs.Reader.Get(ctx, sourceNsScopeKey, sourceNsScope); err != nil {
+			klog.Errorf("Failed to get NSS CR %s: %v, retry again", sourceNsScopeKey.String(), err)
+			continue
+		}
+
+		targetNsScope := &nssv1.NamespaceScope{}
+		targetNsScopeKey := types.NamespacedName{Name: targetCR, Namespace: bs.CSData.MasterNs}
+		if err := bs.Reader.Get(ctx, targetNsScopeKey, targetNsScope); err != nil {
+			klog.Errorf("Failed to get NSS CR %s: %v, retry again", targetNsScopeKey.String(), err)
+			continue
+		}
+
+		sourceNsSet := gset.NewSet()
+		targetNsSet := gset.NewSet()
+		// we cann't convert []T to []interface{} directly in Go, have to add it to set by loop
+		for _, ns := range sourceNsScope.Spec.NamespaceMembers {
+			sourceNsSet.Add(ns)
+		}
+		for _, ns := range targetNsScope.Spec.NamespaceMembers {
+			targetNsSet.Add(ns)
+		}
+
+		// sync up when namepsace in source CR is different from target CR
+		if !sourceNsSet.Equal(targetNsSet) {
+			sourcenNsMems := sourceNsSet.ToSlice()
+			var targetNsMems []string
```

Comment:
Why we create an empty slice here? Could we add `sourcenNsMems` into set `targetNsSet`?

Reply:
The return type of `sourceNsSet.ToSlice()` is `[]interface{}`, but `targetNsScope.Spec.NamespaceMembers` has the type `[]string`, so we have to convert each item to `string` first.

`targetNsMems = append(targetNsMems, ns.(string))`

Reply:
Also, I made some changes to the PR to add the namespaces in `common-service` to `odlm-scope` instead of overwriting it.